Use EmitterProcessor in the FluxMessageChannel

artembilan opened the pull request:
The `EmitterProcessor` has a good logic to block upstream producer
when its downstream subscriber cannot keep up with overproducing.

* Rework `FluxMessageChannel` logic to rely on the `EmitterProcessor`
instead of `Flux.create()`
* Cancel `FluxMessageChannel` internal subscriptions in the `destroy()`
* Fix `ReactiveStreamsTests.testFluxTransform()` for the splitter's
delimiter
* Ensure in the `FluxMessageChannelTests.testFluxMessageChannel`
that we can have several concurrent subscribers to the
`FluxMessageChannel`

<!--
Thanks for contributing to Spring Integration. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc).
-->

artembilan committed:
Use EmitterProcessor in the FluxMessageChannel

The `EmitterProcessor` has a good logic to block upstream producer
when its downstream subscriber cannot keep up with overproducing.

* Rework `FluxMessageChannel` logic to rely on the `EmitterProcessor`
instead of `Flux.create()`
* Cancel `FluxMessageChannel` internal subscriptions in the `destroy()`
* Fix `ReactiveStreamsTests.testFluxTransform()` for the splitter's
delimiter
* Ensure in the `FluxMessageChannelTests.testFluxMessageChannel`
that we can have several concurrent subscribers to the
`FluxMessageChannel`

artembilan commented:
@bsideup ,

Would you mind to take a look into this fix, please?

Thanks

bsideup reviewed (changes requested):
(here was a comment but I missed an important detail so I am updating it :D)

bsideup commented on spring-integration-core/src/main/java/org/springframework/integration/channel/FluxMessageChannel.java:
```diff
@@ -47,15 +49,15 @@
 	private final Map<Publisher<? extends Message<?>>, ConnectableFlux<?>> publishers = new ConcurrentHashMap<>();
 
-	private final Flux<Message<?>> flux;
+	private final Map<ConnectableFlux<?>, Disposable> disposables = new ConcurrentHashMap<>();
```
Consider using `Disposables.composite()`

bsideup commented on spring-integration-core/src/main/java/org/springframework/integration/channel/FluxMessageChannel.java:
```diff
@@ -70,11 +72,12 @@ protected boolean doSend(Message<?> message, long timeout) {
 				.subscribe(subscriber);
 
-		this.publishers.values().forEach(ConnectableFlux::connect);
+		this.publishers.values()
```
could you please explain the use of `ConnectableFlux` here?

artembilan replied:
I need `ConnectableFlux` to subscribe to provideded publishers in lazy-manner: if this `FluxMessageChannel` is subscribed already, I do `connect()` - otherwise I `connect()` to all the provided publishers in its `subscribe()`.
My point is to be able to propagate events from upstream publishers into subscribers of this `FluxMessageChannel` honoring an internal logic of the `MessageChannel` implementation.

I couldn't find any other way in Reactor how to achieve a lazy subscription task unless this `ConnectableFlux`.

And with that it looks like I can't use `Disposables.composite()` since I don't have a way to remove one from that composite when such an upstream published is finished.

I'm OK to change the logic if you can help with that.

Thanks

P.S. Pushed `flux.onComplete()` BTW and removed reference to subscribers since it is not necessary any more.
👍

bsideup commented on spring-integration-core/src/main/java/org/springframework/integration/channel/FluxMessageChannel.java:
```diff
@@ -84,14 +87,22 @@ public void subscribeTo(Publisher<? extends Message<?>> publisher) {
+	public void destroy() {
+		super.destroy();
+
+		this.subscribers.forEach(Subscriber::onComplete);
```
```suggestion
		this.flux.onComplete();
```

artembilan committed:
* Use `flux.onComplete()` instead of iteration over subscribers
* Change `subscribers` list into just `AtomicInteger` count marker
* fix `DefaultSplitterTests` according a new logic in the `FluxMessageChannel`

artembilan committed:
GH-3107: Add errorOnTimeout for TcpInboundGateway

Fixes https://github.com/spring-projects/spring-integration/issues/3107

The `MessagingGatewaySupport` has an `errorOnTimeout` option to throw
a `MessageTimeoutException` when downstream reply doesn't come back in
time for configured reply timeout

* Expose an `errorOnTimeout` option as a `TcpInboundGateway` ctor
property
* Add new factory methods into a `Tcp` factory for Java DSL
* Ensure a property works as expected in the `IpIntegrationTests`
* Document a new option

artembilan commented:
I think I found a solution with the `delaySubscription(Publisher<U> subscriptionDelay)`.

Stay tuned! Will come back to you today 😄

artembilan committed:
* Use `delaySubscription()` for subscribing publishers in the `FluxMessageChannel`
to wait until this one subscribed.
* Use an `EmitterProcessor` to catch subscriptions and pass them as a
signal to delayed upstream publishers
* Fix  `FluxMessageChannelTests.testFluxMessageChannelCleanUp` to
verify an actual property instead of removed.
* Fix `RSocketOutboundGatewayIntegrationTests` for the proper subscription
into a `FluxMessageChannel` before actual interaction with an RSocket
gateway.
This should help us also to avoid some race conditions in the future

artembilan commented:
So, here you are a fix based on the `delaySubscription()` and some additional internal `EmitterProcessor` to handle external subscriptions to this `FluxMessageChannel`.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/channel/FluxMessageChannel.java:
```diff
@@ -43,55 +41,53 @@
+						Mono.fromSupplier(this.subscribed::get)
+								.filter((subscribers) -> subscribers > 0)
+								.switchIfEmpty(this.subscriptionDelay.next()))
+				.subscribe();
```
I think the proper way would be to have a `subscribeTo()` to return `Mono<Void>` instead to let that caller to subscribe or compose further its reactive flow.
This way we would defer an emission from this source publisher as late as possible.

But I'd like to revise that stuff in a separate PR.

Thanks

artembilan committed:
Revert "GH-3107: Add errorOnTimeout for TcpInboundGateway"

This reverts commit fa6119ddc4e423e8b15ca3fdc9fc4c79ea9d56af.

artembilan commented:
@bsideup , gentle ping. 😄

artembilan commented:
OK! I have minimized the change for this PR as much as possible - 2 files. 😄 

Looking forward for your feedback!

artembilan committed:
* Refactor `FluxMessageChannel` to use `ReplayProcessor` for `subscribedSignal`.
This one is used `delaySubscription` for the upstream publishers
* Use a `AtomicBoolean` for subscription state since `doOnSubscribe()`
is called before `EmitterProcessor` adds subscribers for its `downstreams`
* Use `publishOn(Schedulers.boundedElastic())` for upstream publishers
to avoid blocking over there when our `EmitterProcessor` doesn't have
enough demand
* Refactor reactive tests to have a subscription into the `FluxMessageChannel`
earlier than emission happens for it

artembilan commented:
OK. I pushed what we have discussed with some workaround for subscription race condition.

Thank you for your feedback, @bsideup, looking forward for more!

artembilan committed:
* Use `Flux.subscribe(Consumer)` instead of `doOnNext(Consumer).subscribe()`

artembilan commented:
@bsideup ,

so, I have rebased to `master` and decreased a change as much as possible. Again 😄 

Would be great to have some feedback before release this Wednesday.

Thanks

artembilan committed:
* Emit `subscribedSignal` value after `.subscribe(subscriber)`
instead of `doOnSubscribe`
* Check for `this.processor.hasDownstreams()` before emitting such an event

artembilan commented:
Pushed the fix as we discussed in private chat.

Thanks

bsideup commented on spring-integration-core/src/main/java/org/springframework/integration/channel/FluxMessageChannel.java:
```diff
@@ -16,82 +16,81 @@
-
-		this.publishers.values().forEach(ConnectableFlux::connect);
+		if (this.processor.hasDownstreams()) {
+			this.subscribedSignal.onNext(true);
```
probably `this.subscribedSignal.onNext(this.processor.hasDownstreams())` would be better

artembilan replied:
He-he! True. Will fix soon.

artembilan committed:
* Use `this.processor.hasDownstreams()` as a value to emit for `subscribedSignal`.
This way we are less vulnerable race conditions when subscribers are changed
actively

artembilan commented:
@garyrussell ,

We're done here with @bsideup . So, let us know if you are OK and we merge (or you).

Thanks
